sdk/core/core-http/src/fetchHttpClient.ts

```diff
@@ -200,7 +201,23 @@ export abstract class FetchHttpClient implements HttpClient {
     } finally {
       // clean up event listener
       if (httpRequest.abortSignal && abortListener) {
-        httpRequest.abortSignal.removeEventListener("abort", abortListener);
+        let uploadStreamDone = Promise.resolve();
+        if (isReadableStream(body)) {
+          uploadStreamDone = isStreamComplete(body);
+        }
+        let downloadStreamDone = Promise.resolve();
+        if (isReadableStream(operationResponse?.readableStreamBody)) {
+          downloadStreamDone = isStreamComplete(operationResponse!.readableStreamBody);
+        }
+
+        Promise.all([uploadStreamDone, downloadStreamDone])
+          .then(() => {
+            httpRequest.abortSignal?.removeEventListener("abort", abortListener!);
+            return;
+          })
+          .catch((e) => {
+            throw e;
+          });
```

Comment:
I think this is identical, but simpler?

```suggestion
        await Promise.all([uploadStreamDone, downloadStreamDone]);
        httpRequest.abortSignal?.removeEventListener("abort", abortListener!);
```

Reply:
This is exactly what I had first, but for mysterious reason it didn't work and repro program exits silently. Let me try again.

Reply:
Right I had `await` in https://github.com/Azure/azure-sdk-for-js/pull/12038/commits/fc9303621d95e42e03cd0cdcdff7ceea97710a11 and my test program exits silently.  Using callback works.  @bterlson any ideas on the subtle differences between the two that might explain this?

Reply:
Hmm well with the callback method, you're no longer participating in the async method that wraps this call, which means you have an unhandled promise rejection (which in some NodeJS will crash your entire app.)

You don't actually want to do that, I think, since this method is already returning a promise. However, I'm not sure how `await` works inside `finally`... does it override the return/reject state of the promise?

Reply:
Looks like await inside finally *should* do the right thing (throw if there is an error)

Reply:
I did some more experiments. This silent exiting behavior seems to be related to how the helper is implemented. when `await Promise.all()` is used, if I don't have the empty `"data"` event handler the test program exits silently.

```typescript
 function isStreamComplete(stream: Readable): Promise<void> {
   return new Promise((resolve) => {
     stream.on("data", () => {
       return;
     })
     stream.on("close", () => { console.log("### isStreamComplete on close"); resolve(); });
     stream.on("error", () => { console.log("### isStreamComplete on error"); });
     stream.on("end",   () => { console.log("### isStreamComplete on end");   });
   });
 }
```

Also I found that with `await Promise.all()` in our finally block, we are blocking test program's execution until the streaming is finished. User code of `stream.destroy()` on the download stream after `blobClient.download()` call for example won't get executed until download stream is complete, which is wrong.

```
testing 0
### isStreamComplete on end
### isStreamComplete on close
download call done for 0
destroying stream 0...
```

Comparing to the result of using `Promise.all().then()` callback: the stream can be destroyed, and then the promise resolves.

```
testing 0
download call done for 0
destroying stream 0...
### isStreamComplete on end
stream 0 on end
### isStreamComplete on close
stream 0 closed
```

I still don't understand why an empty `"data"` event handler matters. Regardless, we shouldn't block on stream completion.

Reply:
Ah yes, you are absolutely right that we shouldn't block on completion. I think the only dilemma we have here is what to do when the new promise rejects. We can't throw/leave it rejected since the consumer has no way to handle it. 

Should we log or ignore it?

Reply:
Would it be fine if our promise helper always resolves and never rejects? [The doc says](https://nodejs.org/api/stream.html#stream_event_close_1)
> A Readable stream will always emit the 'close' event if it is created with the emitClose option.

Regarding the test program silent exiting when `await Promse.all()` is used. @bterlson helped investigate it and found out that if we await on promise that never resolves or rejects, NodeJS exits the process with [code 13](https://nodejs.org/api/process.html#process_exit_codes)

>*  13 Unfinished Top-Level Await: await was used outside of a function in the top-level code, but the passed Promise never resolved.

Reply:
I'm fine with doing `Promise.all().then(cleanup).catch(() => { /* ignore errors */ })`